Source/CarthageKit/Version.swift

```diff
@@ -221,34 +332,54 @@ public enum VersionSpecifier: VersionType {
 			if let semanticVersion = SemanticVersion.from(version).value {
 				return predicate(semanticVersion)
 			} else {
-				// Consider non-semantic versions (e.g., branches) to meet every
-				// version range requirement.
-				return true
+				// Consider non-semantic versions (e.g., branches) not to meet
+				// any requirement as we can't guarantee any ordering nor
+				// compatibility
+				return false
```

Comment:
We don't want to change this because it's considered a feature: You can specify a branch at the root of your dependency tree and it will override other versions in the tree.

Reply:
Got it. I will update all tests.

Reply:
If I understand correctly, this is actually the original issue that I was trying to solve when I started working on these PR one year ago.

On our repo, someone pushed a tag with pre-release or build metadata. Because it did not parse as a valid version, it will be considered non-sematic. And because of this line you commented on, it would always pick that non semantic one.

If we revert this line, we would go back to that behavior - now pre-releases will be parsed properly, but anyone pushing any other random (non-semantic) tag to the repo would case that tag to always be picked, no matter what is specified in the Cartfile.

E.g. if I specify `>= 5.0.0` in the Cartfile, and the tags on the repo are:

- `0.1.0`
- `Demo`
- `1.0.0`
- ...
- `5.0.1`

It would resolve to `Demo` instead of `5.0.1`. 

It might be the responsibility of the repo mantainers to not allow any tag with non-semantic versioning, but this is something that is easy to overlook. Once one of those has been pushed, it will disrupt everyone working with that repository.

If I understand correctly, the desired behavior is that if I specifiy a commit SHA, a branch name, or a non-sematic tag in the Cartfile in the root project, it will pick that specific commit for this repo and for all dependencies, ignoring whatever the dependency specifies. It makes sense, however this code achieves that with the nasty side effect that I described above. 

Did I get this right?

If this is the case, I would like to implement the desired functionality in a different manner not to have the side effect.

Reply:
> If I understand correctly, this is actually the original issue that I was trying to solve when I started working on these PR one year ago.

Yes and no. This _would_ solve it for tags pre-release builds because those would no longer parse as non-semantic versions. But you're right that a `Demo` tag could still run into the bug you encountered.

The non-semantic version bug really needs to be fixed in the resolver. Even though a non-semantic version is compatible with any version, the resolver shouldn't pick a non-semantic version unless one is specifically requested.

Reply:
My suggestion is then: 
- revert this line to keep the previous behavior
- merge this and the other pre-release PRs to master to mitigate the issue (i.e. at least pre-releases tags will not cause the bug)
- investigate how to fix the resolver so that it does not pick a non-semantic version unless one is specified in a completely separate branch, without having that fix tied to the release of pre-release version parsing

Reply:
That sounds perfect. 👍